### PR TITLE
[Feature] Support K-fold cross-validation

### PR DESCRIPTION
**Pull request by mzr1996**
## Motivation

K-Fold cross-validation is commonly used in small dataset training. Here we add the support of the K-fold dataset wrapper.
Closing #560 

## Modification

1. Add a new dataset wrapper, `KFoldDataset`.
2. Support to use `indices` to specify which samples to evaluate.

## BC-breaking (Optional)

No

## Use cases (Optional)

Here we modify the [resnet18_8xb32_in1k.py](https://github.com/open-mmlab/mmclassification/blob/master/configs/resnet/resnet18_8xb32_in1k.py) as an example.

```python
_base_ = [
    '../_base_/models/resnet18.py',
    '../_base_/schedules/imagenet_bs256.py', '../_base_/default_runtime.py'
]

# dataset settings
dataset_type = 'ImageNet'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='RandomResizedCrop', size=224),
    dict(type='RandomFlip', flip_prob=0.5, direction='horizontal'),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='ToTensor', keys=['gt_label']),
    dict(type='Collect', keys=['img', 'gt_label'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='Resize', size=(256, -1)),
    dict(type='CenterCrop', crop_size=224),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='Collect', keys=['img'])
]


data = dict(
    samples_per_gpu=32,
    workers_per_gpu=2,
    train=dict(
        type='KFoldDataset',
        dataset=dict(
            type=dataset_type,
            data_prefix='data/imagenet/train',
            pipeline=train_pipeline),
        # Modify the `fold` to use different split. For 5-fold cross-validation,
        # five experiments need to be executed (fold=0, fold=1, fold=2, ...)
        fold=0,
        num_splits=5,
        # seed = 1,   # If set seed, shuffle samples before spliting dataset.
    ),
    val=dict(
        type='KFoldDataset',
        dataset=dict(
            type=dataset_type,
            data_prefix='data/imagenet/train',   # For K-Fold cross-validation, all images should be place in the same folder. 
            pipeline=test_pipeline),
        # All parameters need to be the same in train/val/test set, like fold, num_splits and seed
        fold=0,
        num_splits=5,
        # seed=1,
        # `test_mode` will be set to True automatically in `apis/train.py` and `tools/test.py`
        # and you can also specify it explicitly.
        # test_mode=True,
    ),
    test=dict(
        type='KFoldDataset',
        dataset=dict(
            type=dataset_type,
            data_prefix='data/imagenet/train',
            pipeline=test_pipeline),
        fold=0,
        num_splits=5,
        # seed=1,
        # test_mode=True,
    ))
evaluation = dict(interval=1, metric='accuracy')

```

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by liluhu0** [comment]
目前还不支持K折交叉吗？

**Comment by mzr1996** [comment]
> 目前还不支持K折交叉吗？

这个 PR 目前正在开发中，我们会在近期完成支持。
功能本身应该已经可用了，你也可以参照这个 config 试用一下
```python
_base_ = [
    '../_base_/models/resnet18_cifar.py',
    '../_base_/schedules/cifar10_bs128.py', '../_base_/default_runtime.py'
]

# dataset settings
dataset_type = 'CIFAR10'
img_norm_cfg = dict(
    mean=[125.307, 122.961, 113.8575],
    std=[51.5865, 50.847, 51.255],
    to_rgb=False)
train_pipeline = [
    dict(type='RandomCrop', size=32, padding=4),
    dict(type='RandomFlip', flip_prob=0.5, direction='horizontal'),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='ToTensor', keys=['gt_label']),
    dict(type='Collect', keys=['img', 'gt_label'])
]
test_pipeline = [
    dict(type='Normalize', **img_norm_cfg),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='Collect', keys=['img'])
]
data = dict(
    samples_per_gpu=16,
    workers_per_gpu=2,
    train=dict(
        type='KFoldDataset',
        dataset=dict(
            type=dataset_type,
            data_prefix='data/cifar10',
            pipeline=train_pipeline),
        fold=0,
        num_splits=5),
    val=dict(
        type='KFoldDataset',
        dataset=dict(
            type=dataset_type,
            data_prefix='data/cifar10',
            pipeline=test_pipeline),
        fold=0,
        num_splits=5),
    test=dict(
        type='KFoldDataset',
        dataset=dict(
            type=dataset_type,
            data_prefix='data/cifar10',
            pipeline=test_pipeline),
        fold=0,
        num_splits=5))
```

File: mmcls/datasets/dataset_wrappers.py [hunk 1]
```diff
@@ -170,3 +170,41 @@ def __getitem__(self, idx):
...
+        self.num_splits = num_splits
+
+        length = len(dataset)
+        indices = list(range(length))
+        if isinstance(seed, int):
+            rng = np.random.default_rng(seed)
+            rng.shuffle(indices)
+
+        test_start = length // num_splits * fold
+        test_end = length // num_splits * (fold + 1)
```
Comment (0x4f5da2) [inline mmcls/datasets/dataset_wrappers.py:196]: Current implementation may omit some samples due to rounding.
 
Let's say we have 29 samples split into 5 folds. For the last fold, `test_start = 20` and `test_end = 25`.


```suggestion
        test_start = length * fold // num_splits
        test_end = length * (fold + 1) // num_splits
```

Comment (mzr1996) [reply mmcls/datasets/dataset_wrappers.py]: Done

File: tools/kfold-cross-valid.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+TEST_METRICS = ('precision', 'recall', 'f1_score', 'support', 'mAP', 'CP',
+                'CR', 'CF1', 'OP', 'OR', 'OF1', 'accuracy')
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Train a model')
+    parser.add_argument('config', help='train config file path')
+    parser.add_argument('--num-splits', type=int, help='TODO')
+    parser.add_argument('--fold', type=int, help='TODO')
+    parser.add_argument('--summary', action='store_true', help='TODO')
```
Comment (0x4f5da2) [inline tools/kfold-cross-valid.py:31]: replace todo with illustration.

**Review by 0x4f5da2 — APPROVED** [review]

**Pull request merged**